app/decorators/spree/core/product_duplicator_decorator.rb

```diff
@@ -0,0 +1,24 @@
+module Spree
+  module Core
+    module ProductDuplicatorDecorator
+      def duplicate
+        new_product = super
+
+        product.videos.each do |video|
+          new_video = video.dup
+          new_video.product_id = new_product.id
+          new_video.save!
+
+          video.tags.each do |tag|
+            video_tag = Spree::ProductVideoTag.find_by(name: tag.name)
+            new_video.product_video_taggings.create!(product_video_tag_id: video_tag.id)
+          end
+        end
+
+        new_product
+      end
+    end
+  end
+end
+
+Spree::ProductDuplicator.prepend Spree::Core::ProductDuplicatorDecorator
```

Comment:
I can see N+1 queries in the code, instead lets rewrite

```
module Spree
  module Core
    module ProductDuplicatorDecorator
      def duplicate
        ActiveRecord::Base.transaction do
          new_product = super
          duplicate_videos(new_product)
          new_product
        end
      end

      private

      def duplicate_videos(new_product)
        video_data = product.videos.includes(:tags).map do |video|
          new_video = video.dup
          new_video.product_id = new_product.id
          new_video.tag_ids = video.tags.pluck(:id)
          new_video
        end

        Spree::Video.import(video_data, recursive: true)
      end
    end
  end
end

Spree::ProductDuplicator.prepend Spree::Core::ProductDuplicatorDecorator
```

To make this work, you'll need to ensure a few things:

Make sure you have the activerecord-import gem in your Gemfile and installed.
Ensure that your Spree::Video model has a tag_ids= method, which is typically provided by a has_many :tags or has_and_belongs_to_many :tags association.
If you're using has_many :through for the tags association, you might need to adjust the tag_ids assignment slightly.